File: clients/tui/sdk/python/tui2sdk/wire.py
"""TUI v2 wire codec (stdlib only): protobuf payloads + binary frames.

This is the Python binding of ``proto/ui/tui2.proto`` (the single source of
truth) for the subset a layout program needs:

  decode: HELLO, EVENT (key/paste/mouse/wheel/resize/sources/notice/
          component/view_rejected), RESPONSE
  encode: VIEW (Box tree, Keys), RESULT (MethodParams)

Wire basics: every field is ``varint(tag) | payload`` with
``tag = (field << 3) | wire_type``; 0 = varint, 1 = 64-bit, 2 =
length-delimited, 5 = 32-bit. A frame is
``u32 length (big endian) | u8 type | payload`` where length counts the type
byte plus the payload (PROTOCOL §0).

The codec deliberately depends on nothing outside the standard library: any
language can be a layout program, this module is only the Python convenience
binding.
"""

import struct

# Frame types (PROTOCOL §0): fixed, append-only.
HELLO = 1
VIEW = 2
EVENT = 3
RESULT = 4
RESPONSE = 5

WIRE_VARINT = 0
WIRE_64BIT = 1
WIRE_LEN = 2
WIRE_32BIT = 5

MAX_MESSAGE_BYTES = 1 << 20


class WireError(Exception):
    """Raised for malformed frames, payloads or protocol misuse."""


# ----------------------------------------------------------------- encoding

def uvarint(value):
    """Encode an int as an unsigned varint. Negative int32/int64 values are
    sign-extended to 64 bits, exactly like protobuf does."""
    if value < 0:
        value += 1 << 64
    out = bytearray()
    while True:
        byte = value & 0x7F
        value >>= 7
        if value:
            out.append(byte | 0x80)
        else:
            out.append(byte)
            return bytes(out)


def field_varint(field, value):
    if value is None:
        return b""
    return uvarint(field << 3) + uvarint(int(value))


def field_bool(field, value):
    if value is None:
        return b""
    return field_varint(field, 1 if value else 0)


def field_bytes(field, data):
    if not data:
        return b""
    return uvarint((field << 3) | WIRE_LEN) + uvarint(len(data)) + data


def field_string(field, text):
    if not text:
        return b""
    return field_bytes(field, text.encode("utf-8"))


def field_msg(field, message):
    if message is None:
        return b""
    return field_bytes(field, message)


def field_map_string_string(field, mapping):
    out = b""
    for key in sorted(mapping):
        entry = field_string(1, key) + field_string(2, mapping[key])
        out += field_msg(field, entry)
    return out


# ----------------------------------------------------------------- decoding

def read_uvarint(buf, index):
    result = 0
    shift = 0
    while True:
        byte = buf[index]
        index += 1
        result |= (byte & 0x7F) << shift
        if not byte & 0x80:
            return result, index
        shift += 7


def parse_fields(data):
    """Parse one message body into [(field, wire, value), ...] preserving
    order. Varint values are ints, length-delimited values are bytes."""
    out = []
    index = 0
    size = len(data)
    while index < size:
        tag, index = read_uvarint(data, index)
        field, wire = tag >> 3, tag & 7
        if wire == WIRE_VARINT:
            value, index = read_uvarint(data, index)
        elif wire == WIRE_LEN:
            length, index = read_uvarint(data, index)
            value = data[index:index + length]
            index += length
        elif wire == WIRE_64BIT:
            value = data[index:index + 8]
            index += 8
        elif wire == WIRE_32BIT:
            value = data[index:index + 4]
            index += 4
        else:
            raise WireError("unsupported wire type %d" % wire)
        out.append((field, wire, value))
    return out


def to_int32(value):
    return value - (1 << 32) if value >= 1 << 31 else value


def to_int64(value):
    return value - (1 << 64) if value >= 1 << 63 else value


def _first(fields, field):
    for number, _wire, value in fields:
        if number == field:
            return value
    return None


def _strings(fields, field):
    return [value.decode("utf-8") for number, _wire, value in fields if number == field]


def _string(fields, field):
    value = _first(fields, field)
    return value.decode("utf-8") if value is not None else ""


def _uint(fields, field):
    value = _first(fields, field)
    return value if value is not None else 0


def _bool(fields, field):
    return _uint(fields, field) != 0


def parse_map_string_bool(entry):
    fields = parse_fields(entry)
    return {_string(fields, 1): _bool(fields, 2)}


def parse_limits(data):
    fields = parse_fields(data)
    return {
        "max_nodes": _uint(fields, 1),
        "max_message_bytes": _uint(fields, 2),
        "max_paste_bytes": _uint(fields, 3),
        "max_inflight_requests": _uint(fields, 4),
        "owner_lease_ttl_ms": _uint(fields, 5),
    }


def decode_hello(data):
    fields = parse_fields(data)
    hello = {
        "schema": _uint(fields, 1),
        "view_id": _string(fields, 2),
        "epoch": _uint(fields, 3),
        "cols": _uint(fields, 4),
        "rows": _uint(fields, 5),
        "components": _strings(fields, 6),
        "events": _strings(fields, 7),
        "methods": _strings(fields, 8),
        "features": {},
        "limits": {},
    }
    for field, wire, value in fields:
        if field == 9 and wire == WIRE_LEN:
            hello["features"].update(parse_map_string_bool(value))
        elif field == 10 and wire == WIRE_LEN:
            hello["limits"] = parse_limits(value)
    return hello


def encode_hello(hello):
    out = field_varint(1, hello.get("schema"))
    out += field_string(2, hello.get("view_id"))
    out += field_varint(3, hello.get("epoch"))
    out += field_varint(4, hello.get("cols"))
    out += field_varint(5, hello.get("rows"))
    for name in hello.get("components", ()):
        out += field_string(6, name)
    for name in hello.get("events", ()):
        out += field_string(7, name)
    for name in hello.get("methods", ()):
        out += field_string(8, name)
    if hello.get("features"):
        out += field_map_string_bool(9, hello["features"])
    if hello.get("limits"):
        out += field_msg(10, encode_limits(hello["limits"]))
    return out


def encode_limits(limits):
    out = field_varint(1, limits.get("max_nodes"))
    out += field_varint(2, limits.get("max_message_bytes"))
    out += field_varint(3, limits.get("max_paste_bytes"))
    out += field_varint(4, limits.get("max_inflight_requests"))
    out += field_varint(5, limits.get("owner_lease_ttl_ms"))
    return out


def field_map_string_bool(field, mapping):
    out = b""
    for key in sorted(mapping):
        entry = field_string(1, key) + field_bool(2, mapping[key])
        out += field_msg(field, entry)
    return out


def parse_source(data):
    fields = parse_fields(data)
    return {
        "id": _string(fields, 1),
        "kind": _string(fields, 2),
        "title": _string(fields, 3),
        "endpoint": _string(fields, 4),
        "terminal_id": _string(fields, 5),
        "attached": _bool(fields, 6),
        "exited": _bool(fields, 7),
        "exit_code": to_int32(_uint(fields, 8)),
        "health": _string(fields, 9),
        "resize_owner": _string(fields, 10),
        "owner_epoch": _uint(fields, 11),
        "last_seen_ms": to_int64(_uint(fields, 12)),
    }


def decode_event(data):
    """Decode one EVENT payload into {"kind": ..., plus the payload keys}."""
    for field, wire, value in parse_fields(data):
        if wire != WIRE_LEN:
            continue
        inner = parse_fields(value)
        if field == 1:  # key
            return {"kind": "key", "id": _string(inner, 1),
                    "key": _string(inner, 2), "char": _string(inner, 3)}
        if field == 2:  # paste
            return {"kind": "paste", "id": _string(inner, 1), "text": _string(inner, 2)}
        if field == 3:  # mouse
            return {"kind": "mouse", "action": _string(inner, 1), "button": _string(inner, 2),
                    "x": to_int32(_uint(inner, 3)), "y": to_int32(_uint(inner, 4)),
                    "node": _string(inner, 5)}
        if field == 4:  # wheel
            return {"kind": "wheel", "delta": to_int32(_uint(inner, 1)),
                    "x": to_int32(_uint(inner, 2)), "y": to_int32(_uint(inner, 3)),
                    "node": _string(inner, 4)}
        if field == 5:  # resize
            return {"kind": "resize", "cols": _uint(inner, 1), "rows": _uint(inner, 2)}
        if field == 6:  # sources
            items = []
            for number, item_wire, item in inner:
                if number == 1 and item_wire == WIRE_LEN:
                    items.append(parse_source(item))
            return {"kind": "sources", "items": items}
        if field == 7:  # notice
            return {"kind": "notice", "level": _string(inner, 1), "message": _string(inner, 2)}
        if field == 8:  # component
            return {"kind": "component", "source": _string(inner, 1),
                    "name": _string(inner, 2), "value": _string(inner, 3)}
        if field == 9:  # view_rejected
            return {"kind": "view_rejected", "epoch": _uint(inner, 1),
                    "rev": _uint(inner, 2), "reason": _string(inner, 3)}
    return {"kind": "unknown"}


def decode_response(data):
    fields = parse_fields(data)
    response = {
        "request_id": _uint(fields, 1),
        "epoch": _uint(fields, 2),
        "ok": _bool(fields, 3),
        "data": {},
        "error": _string(fields, 5),
    }
    for field, wire, value in fields:
        if field == 4 and wire == WIRE_LEN:
            inner = parse_fields(value)
            response["data"] = {
                "rows": _strings(inner, 1),
                "text": _string(inner, 2),
                "endpoint": _string(inner, 3),
                "id": _string(inner, 4),
            }
    return response


# ---------------------------------------------------------------- box trees

def encode_cursor(cursor):
    """Encode a program cursor dict {row,col,shape?,visible?} (Box field 7).

    ``visible`` is an optional proto3 field: it is only written when the
    program sets it, so absence keeps the kernel default (visible).
    """
    row, col = cursor.get("row", 0), cursor.get("col", 0)
    out = field_varint(1, row) + field_varint(2, col)
    out += field_string(3, cursor.get("shape"))
    if cursor.get("visible") is not None:
        out += field_bool(4, cursor["visible"])
    return out


def encode_box(box):
    """Encode one view-tree node dict (keys mirror tui2.proto Box)."""
    out = field_string(1, box.get("id"))
    if box.get("size"):
        width, height, flex = box["size"]
        out += field_msg(2, field_varint(1, width) + field_varint(2, height) + field_varint(3, flex))
    if box.get("pos") is not None:
        x, y = box["pos"]
        out += field_msg(3, field_varint(1, x) + field_varint(2, y))
    out += field_string(4, box.get("flow"))
    if box.get("visible") is not None:
        out += field_bool(5, box["visible"])
    if box.get("cursor") is not None:
        out += field_msg(7, encode_cursor(box["cursor"]))
    content = box.get("content")
    if content:
        encoded = field_string(1, content.get("text"))
        for line in content.get("lines", ()):
            encoded += field_string(2, line)
        encoded += field_string(3, content.get("self"))
        if content.get("props"):
            encoded += field_map_string_string(4, content["props"])
        out += field_msg(8, encoded)
    for kind in box.get("input", ()):
        out += field_string(9, kind)
    out += field_bool(10, box.get("focused"))
    for child in box.get("children", ()):
        out += field_msg(11, encode_box(child))
    out += field_string(12, box.get("style"))
    return out


def encode_view(epoch, rev, claim, all_keys, root):
    keys = b""
    for key in claim:
        keys += field_string(1, key)
    keys += field_bool(2, all_keys)
    return (field_varint(1, epoch) + field_varint(2, rev) +
            field_msg(3, keys) + field_msg(4, encode_box(root)))


# ------------------------------------------------------------------- result

def encode_result(request_id, epoch, method, params=None):
    params = params or {}
    encoded = field_string(1, params.get("endpoint"))
    encoded += field_string(2, params.get("id"))
    if params.get("fit") is not None:
        encoded += field_bool(3, params["fit"])
    if params.get("expected_owner_epoch") is not None:
        encoded += field_varint(4, params["expected_owner_epoch"])
    for arg in params.get("argv", ()):
        encoded += field_string(5, arg)
    encoded += field_string(6, params.get("cwd"))
    if params.get("env"):
        encoded += field_map_string_string(7, params["env"])
    encoded += field_string(8, params.get("title"))
    if params.get("ephemeral") is not None:
        encoded += field_bool(9, params["ephemeral"])
    if params.get("delta") is not None:
        encoded += field_varint(10, params["delta"])
    if params.get("sel") is not None:
        mode, start, end = params["sel"]
        encoded += field_msg(11, field_string(1, mode) + field_varint(2, start) + field_varint(3, end))
    if params.get("offset") is not None:
        encoded += field_varint(12, params["offset"])
    if params.get("rows") is not None:
        encoded += field_varint(13, params["rows"])
    encoded += field_string(14, params.get("event_id"))
    encoded += field_string(15, params.get("source"))
    if params.get("cleanup_owned") is not None:
        encoded += field_bool(16, params["cleanup_owned"])
    # Endpoint connection metadata (tui2.proto MethodParams fields 17..20):
    # kind "command"/"daemon", socket path, connect_mode and the optional
    # tcp address (HOST:PORT).
    encoded += field_string(17, params.get("kind"))
    encoded += field_string(18, params.get("socket"))
    encoded += field_string(19, params.get("connect_mode"))
    encoded += field_string(20, params.get("address"))
    return field_varint(1, request_id) + field_varint(2, epoch) + \
        field_string(3, method) + field_msg(4, encoded)


# -------------------------------------------------------------------- frame

def frame(frame_type, payload):
    """One envelope: u32 length (type + payload, big endian) | u8 type | body."""
    body = bytes([frame_type]) + payload
    return struct.pack(">I", len(body)) + body


def read_frame(stream):
    """Read one frame, returning (type, payload) or None on clean EOF."""
    prefix = stream.read(4)
    if not prefix:
        return None
    if len(prefix) < 4:
        raise WireError("truncated frame length")
    (length,) = struct.unpack(">I", prefix)
    if length == 0:
        raise WireError("zero-length frame")
    if length > MAX_MESSAGE_BYTES:
        raise WireError("frame exceeds max_message_bytes")
    body = stream.read(length)
    if len(body) < length:
        raise WireError("truncated frame")
    return body[0], body[1:]

File: clients/tui/sdk/ts/src/wire.js
/**
 * TUI v2 wire codec (Node stdlib only): protobuf payloads + binary frames.
 * The TypeScript-facing types live in ../../index.d.ts.
 *
 * Wire basics: `varint(tag) | payload`, tag = (field << 3) | wire_type;
 * 0 = varint, 1 = 64-bit, 2 = length-delimited, 5 = 32-bit. A frame is
 * `u32 length (big-endian) | u8 type | payload` (PROTOCOL §0).
 */

'use strict';

const HELLO = 1;
const VIEW = 2;
const EVENT = 3;
const RESULT = 4;
const RESPONSE = 5;

const WIRE_VARINT = 0;
const WIRE_64BIT = 1;
const WIRE_LEN = 2;
const WIRE_32BIT = 5;

const MAX_MESSAGE_BYTES = 1 << 20;

class WireError extends Error {}

// ----------------------------------------------------------------- encoding

function uvarint(value) {
  let big = BigInt(value);
  if (big < 0n) big += 1n << 64n;
  const out = [];
  for (;;) {
    const byte = Number(big & 0x7fn);
    big >>= 7n;
    if (big) out.push(byte | 0x80);
    else {
      out.push(byte);
      return Buffer.from(out);
    }
  }
}

function fieldVarint(field, value) {
  if (value === undefined || value === null) return Buffer.alloc(0);
  return Buffer.concat([uvarint(field << 3), uvarint(value)]);
}

function fieldBool(field, value) {
  if (value === undefined || value === null) return Buffer.alloc(0);
  return fieldVarint(field, value ? 1 : 0);
}

function fieldBytes(field, data) {
  if (!data || data.length === 0) return Buffer.alloc(0);
  return Buffer.concat([uvarint((field << 3) | WIRE_LEN), uvarint(data.length), Buffer.from(data)]);
}

function fieldString(field, text) {
  if (!text) return Buffer.alloc(0);
  return fieldBytes(field, Buffer.from(text, 'utf8'));
}

function fieldMsg(field, message) {
  if (message === null || message === undefined) return Buffer.alloc(0);
  return fieldBytes(field, message);
}

function fieldMapStringString(field, mapping) {
  const parts = [];
  for (const key of Object.keys(mapping).sort()) {
    parts.push(Buffer.concat([fieldString(1, key), fieldString(2, mapping[key])]));
  }
  return parts.map((entry) => fieldMsg(field, entry)).reduce((a, b) => Buffer.concat([a, b]), Buffer.alloc(0));
}

function fieldMapStringBool(field, mapping) {
  const parts = [];
  for (const key of Object.keys(mapping).sort()) {
    parts.push(Buffer.concat([fieldString(1, key), fieldBool(2, mapping[key])]));
  }
  return parts.map((entry) => fieldMsg(field, entry)).reduce((a, b) => Buffer.concat([a, b]), Buffer.alloc(0));
}

// ----------------------------------------------------------------- decoding

function readUvarint(buf, index) {
  let result = 0n;
  let shift = 0n;
  for (;;) {
    const byte = buf[index];
    index += 1;
    result |= BigInt(byte & 0x7f) << shift;
    if (!(byte & 0x80)) return [result, index];
    shift += 7n;
  }
}

function parseFields(data) {
  const out = [];
  let index = 0;
  while (index < data.length) {
    let tag;
    [tag, index] = readUvarint(data, index);
    const field = Number(tag >> 3n);
    const wire = Number(tag & 7n);
    let value;
    if (wire === WIRE_VARINT) {
      [value, index] = readUvarint(data, index);
    } else if (wire === WIRE_LEN) {
      let length;
      [length, index] = readUvarint(data, index);
      length = Number(length);
      value = data.subarray(index, index + length);
      index += length;
    } else if (wire === WIRE_64BIT) {
      value = data.subarray(index, index + 8);
      index += 8;
    } else if (wire === WIRE_32BIT) {
      value = data.subarray(index, index + 4);
      index += 4;
    } else {
      throw new WireError(`unsupported wire type ${wire}`);
    }
    out.push([field, wire, value]);
  }
  return out;
}

function toInt32(value) {
  const big = BigInt(value);
  return Number(BigInt.asIntN(32, big));
}

function toInt64(value) {
  const big = BigInt(value);
  return Number(BigInt.asIntN(64, big));
}

function first(fields, field) {
  for (const [number, , value] of fields) if (number === field) return value;
  return undefined;
}

function strings(fields, field) {
  return fields.filter(([number]) => number === field).map(([, , value]) => value.toString('utf8'));
}

function string(fields, field) {
  const value = first(fields, field);
  return value === undefined ? '' : value.toString('utf8');
}

function uint(fields, field) {
  const value = first(fields, field);
  return value === undefined ? 0 : Number(value);
}

function bool(fields, field) {
  return uint(fields, field) !== 0;
}

function parseLimits(data) {
  const fields = parseFields(data);
  return {
    max_nodes: uint(fields, 1),
    max_message_bytes: uint(fields, 2),
    max_paste_bytes: uint(fields, 3),
    max_inflight_requests: uint(fields, 4),
    owner_lease_ttl_ms: uint(fields, 5),
  };
}

function parseMapStringBool(entry) {
  const fields = parseFields(entry);
  return { [string(fields, 1)]: bool(fields, 2) };
}

function decodeHello(data) {
  const fields = parseFields(data);
  const hello = {
    schema: uint(fields, 1), view_id: string(fields, 2), epoch: uint(fields, 3),
    cols: uint(fields, 4), rows: uint(fields, 5),
    components: strings(fields, 6), events: strings(fields, 7), methods: strings(fields, 8),
    features: {}, limits: {},
  };
  for (const [field, wire, value] of fields) {
    if (field === 9 && wire === WIRE_LEN) Object.assign(hello.features, parseMapStringBool(value));
    else if (field === 10 && wire === WIRE_LEN) hello.limits = parseLimits(value);
  }
  return hello;
}

function parseSource(data) {
  const fields = parseFields(data);
  return {
    id: string(fields, 1), kind: string(fields, 2), title: string(fields, 3),
    endpoint: string(fields, 4), terminal_id: string(fields, 5),
    attached: bool(fields, 6), exited: bool(fields, 7),
    exit_code: toInt32(uint(fields, 8)), health: string(fields, 9),
    resize_owner: string(fields, 10), owner_epoch: uint(fields, 11),
    last_seen_ms: toInt64(uint(fields, 12)),
  };
}

function decodeEvent(data) {
  for (const [field, wire, value] of parseFields(data)) {
    if (wire !== WIRE_LEN) continue;
    const inner = parseFields(value);
    if (field === 1) return { kind: 'key', id: string(inner, 1), key: string(inner, 2), char: string(inner, 3) };
    if (field === 2) return { kind: 'paste', id: string(inner, 1), text: string(inner, 2) };
    if (field === 3) return {
      kind: 'mouse', action: string(inner, 1), button: string(inner, 2),
      x: toInt32(uint(inner, 3)), y: toInt32(uint(inner, 4)), node: string(inner, 5),
    };
    if (field === 4) return {
      kind: 'wheel', delta: toInt32(uint(inner, 1)), x: toInt32(uint(inner, 2)),
      y: toInt32(uint(inner, 3)), node: string(inner, 4),
    };
    if (field === 5) return { kind: 'resize', cols: uint(inner, 1), rows: uint(inner, 2) };
    if (field === 6) {
      const items = [];
      for (const [number, itemWire, item] of inner) {
        if (number === 1 && itemWire === WIRE_LEN) items.push(parseSource(item));
      }
      return { kind: 'sources', items };
    }
    if (field === 7) return { kind: 'notice', level: string(inner, 1), message: string(inner, 2) };
    if (field === 8) return { kind: 'component', source: string(inner, 1), name: string(inner, 2), value: string(inner, 3) };
    if (field === 9) return { kind: 'view_rejected', epoch: uint(inner, 1), rev: uint(inner, 2), reason: string(inner, 3) };
  }
  return { kind: 'unknown' };
}

function decodeResponse(data) {
  const fields = parseFields(data);
  const response = {
    request_id: uint(fields, 1), epoch: uint(fields, 2), ok: bool(fields, 3),
    data: {}, error: string(fields, 5),
  };
  for (const [field, wire, value] of fields) {
    if (field === 4 && wire === WIRE_LEN) {
      const inner = parseFields(value);
      response.data = {
        rows: strings(inner, 1), text: string(inner, 2),
        endpoint: string(inner, 3), id: string(inner, 4),
      };
    }
  }
  return response;
}

// ---------------------------------------------------------------- box trees

function encodeCursor(cursor) {
  let out = Buffer.concat([fieldVarint(1, cursor.row || 0), fieldVarint(2, cursor.col || 0)]);
  out = Buffer.concat([out, fieldString(3, cursor.shape)]);
  if (cursor.visible !== undefined && cursor.visible !== null) out = Buffer.concat([out, fieldBool(4, cursor.visible)]);
  return out;
}

function encodeBox(box) {
  let out = fieldString(1, box.id);
  if (box.size) {
    const size = Buffer.concat([
      fieldVarint(1, box.size[0]),
      fieldVarint(2, box.size[1]),
      fieldVarint(3, box.size[2] || 0),
    ]);
    out = Buffer.concat([out, fieldMsg(2, size)]);
  }
  if (box.pos) {
    const pos = Buffer.concat([fieldVarint(1, box.pos[0]), fieldVarint(2, box.pos[1])]);
    out = Buffer.concat([out, fieldMsg(3, pos)]);
  }
  out = Buffer.concat([out, fieldString(4, box.flow)]);
  if (box.visible !== undefined && box.visible !== null) out = Buffer.concat([out, fieldBool(5, box.visible)]);
  if (box.cursor) out = Buffer.concat([out, fieldMsg(7, encodeCursor(box.cursor))]);
  if (box.content) {
    const content = box.content;
    let encoded = fieldString(1, content.text);
    for (const line of content.lines || []) encoded = Buffer.concat([encoded, fieldString(2, line)]);
    encoded = Buffer.concat([encoded, fieldString(3, content.self)]);
    if (content.props && Object.keys(content.props).length) encoded = Buffer.concat([encoded, fieldMapStringString(4, content.props)]);
    out = Buffer.concat([out, fieldMsg(8, encoded)]);
  }
  for (const kind of box.input || []) out = Buffer.concat([out, fieldString(9, kind)]);
  out = Buffer.concat([out, fieldBool(10, box.focused)]);
  for (const child of box.children || []) out = Buffer.concat([out, fieldMsg(11, encodeBox(child))]);
  return Buffer.concat([out, fieldString(12, box.style)]);
}

function encodeView(epoch, rev, claim, allKeys, root) {
  let keys = Buffer.alloc(0);
  for (const key of claim) keys = Buffer.concat([keys, fieldString(1, key)]);
  keys = Buffer.concat([keys, fieldBool(2, allKeys)]);
  return Buffer.concat([
    fieldVarint(1, epoch), fieldVarint(2, rev), fieldMsg(3, keys), fieldMsg(4, encodeBox(root))]);
}

// ------------------------------------------------------------------- result

function encodeResult(requestId, epoch, method, params) {
  const p = params || {};
  let encoded = fieldString(1, p.endpoint);
  encoded = Buffer.concat([encoded, fieldString(2, p.id)]);
  if (p.fit !== undefined && p.fit !== null) encoded = Buffer.concat([encoded, fieldBool(3, p.fit)]);
  if (p.expected_owner_epoch !== undefined && p.expected_owner_epoch !== null) {
    encoded = Buffer.concat([encoded, fieldVarint(4, p.expected_owner_epoch)]);
  }
  for (const arg of p.argv || []) encoded = Buffer.concat([encoded, fieldString(5, arg)]);
  encoded = Buffer.concat([encoded, fieldString(6, p.cwd)]);
  if (p.env && Object.keys(p.env).length) encoded = Buffer.concat([encoded, fieldMapStringString(7, p.env)]);
  encoded = Buffer.concat([encoded, fieldString(8, p.title)]);
  if (p.ephemeral !== undefined && p.ephemeral !== null) encoded = Buffer.concat([encoded, fieldBool(9, p.ephemeral)]);
  if (p.delta !== undefined && p.delta !== null) encoded = Buffer.concat([encoded, fieldVarint(10, p.delta)]);
  if (p.sel) encoded = Buffer.concat([encoded, fieldMsg(11, Buffer.concat([
    fieldString(1, p.sel.mode), fieldVarint(2, p.sel.start), fieldVarint(3, p.sel.end)]))]);
  if (p.offset !== undefined && p.offset !== null) encoded = Buffer.concat([encoded, fieldVarint(12, p.offset)]);
  if (p.rows !== undefined && p.rows !== null) encoded = Buffer.concat([encoded, fieldVarint(13, p.rows)]);
  encoded = Buffer.concat([encoded, fieldString(14, p.event_id)]);
  encoded = Buffer.concat([encoded, fieldString(15, p.source)]);
  if (p.cleanup_owned !== undefined && p.cleanup_owned !== null) encoded = Buffer.concat([encoded, fieldBool(16, p.cleanup_owned)]);
  encoded = Buffer.concat([encoded, fieldString(17, p.kind)]);
  encoded = Buffer.concat([encoded, fieldString(18, p.socket)]);
  encoded = Buffer.concat([encoded, fieldString(19, p.connect_mode)]);
  encoded = Buffer.concat([encoded, fieldString(20, p.address)]);
  return Buffer.concat([
    fieldVarint(1, requestId), fieldVarint(2, epoch), fieldString(3, method), fieldMsg(4, encoded)]);
}

// -------------------------------------------------------------------- frame

function frame(frameType, payload) {
  const body = Buffer.concat([Buffer.from([frameType]), payload]);
  const prefix = Buffer.alloc(4);
  prefix.writeUInt32BE(body.length, 0);
  return Buffer.concat([prefix, body]);
}

/** Async reader: `next()` resolves `{type, payload}` or null on clean EOF. */
class FrameReader {
  constructor(stream) {
    this.buffer = Buffer.alloc(0);
    this.waiters = [];
    this.done = false;
    this.error = null;
    stream.on('data', (chunk) => {
      this.buffer = Buffer.concat([this.buffer, chunk]);
      this.pump();
    });
    stream.on('end', () => {
      this.done = true;
      this.pump();
    });
    stream.on('error', (error) => {
      this.error = error;
      this.pump();
    });
  }

  next() {
    return new Promise((resolve, reject) => {
      this.waiters.push({ resolve, reject });
      this.pump();
    });
  }

  pump() {
    while (this.waiters.length) {
      let parsed;
      try {
        parsed = this.tryRead();
      } catch (error) {
        const waiter = this.waiters.shift();
        waiter.reject(error);
        continue;
      }
      if (parsed === undefined) return;
      const waiter = this.waiters.shift();
      waiter.resolve(parsed);
    }
  }

  tryRead() {
    if (this.error) throw this.error;
    if (this.buffer.length < 4) return this.done ? null : undefined;
    const length = this.buffer.readUInt32BE(0);
    if (length === 0) throw new WireError('zero-length frame');
    if (length > MAX_MESSAGE_BYTES) throw new WireError('frame exceeds max_message_bytes');
    if (this.buffer.length < 4 + length) {
      if (this.done) throw new WireError('truncated frame');
      return undefined;
    }
    const body = this.buffer.subarray(4, 4 + length);
    this.buffer = this.buffer.subarray(4 + length);
    return { type: body[0], payload: body.subarray(1) };
  }
}

module.exports = {
  HELLO, VIEW, EVENT, RESULT, RESPONSE, MAX_MESSAGE_BYTES, WireError,
  uvarint, fieldVarint, fieldBool, fieldBytes, fieldString, fieldMsg,
  fieldMapStringString, fieldMapStringBool, parseFields, readUvarint,
  toInt32, toInt64, decodeHello, decodeEvent, decodeResponse,
  encodeBox, encodeCursor, encodeView, encodeResult, frame, FrameReader,
};
